packages/query-core/src/tests/queryClient.test.tsx

```diff
@@ -357,6 +358,38 @@ describe('queryClient', () => {
       expect(queryCache.find(key)!.state.data).toBe(newData)
     })
 
+    test('should apply a custom structuralSharing function when provided', () => {
+      const key = queryKey()
+
+      const queryObserverOptions = {
+        structuralSharing: (
+          prevData: { value: Date } | undefined,
+          newData: { value: Date },
+        ) => {
+          if (!prevData) {
+            return newData
+          }
+          return newData.value.getTime() === prevData.value.getTime()
+            ? prevData
+            : newData
+        },
+      } as QueryObserverOptions
```

Comment:
It's not clear to me how I could avoid type casting here, since `setDefaultOptions` takes `DefaultOptions`, which has `QueryObserverOptions<unknown, TError>` as the type of `query`, which isn't going to match any function which accepts a concrete type for `TData`. I think you have exactly the same problem if you try to update `onSuccess` this way, so the problem isn't limited to the change introduced in this PR.

Reply:
can you just inline the object? like:

```
queryClient.setDefaultOptions({ queries: { structuralSharing: () ... } })
```

or use a const assertion: `as const` instead of `as QueryObserverOptions`

Reply:
will you still address this @richsilv ?

Reply:
Sorry for the delay, fixing this up now.

Reply:
No, I'm afraid that neither of those will work. The problem is that `setDefaultOptions` is typed to take a concrete (non-generic) type, `DefaultOptions`, which has a concrete sub-type of `QueryObserverOptions` for it's `query` field with `TData` set to `unknown`. So I don't think there will be any way to make it type check when passed a callback for a query which has a non-`unknown` type for `TData` without asserting like this, as TS will always complain that the function you're providing expects something other than `unknown` for its data argument. The only other alternative is to change the typing for `DefaultOptions`, which I'm happy to try but I suspect will be a much larger diff.

As I say, this isn't restricted to the new `structuralSharing` field, it also applies to `onSuccess`.

Reply:
ah yeah, if you use defaults, there is no way to know which value will be provided for data, because it can and will be used for multiple queries with different structures.

I'm not really sure why the `as QueryObserverOptions` solves it, because that one is also bound to `unknown` but 🤷

Reply:
I think it's because by asserting as a type that's bound to `unknown` you're pretending that `TData` is `unknown` and therefore consistent with `DefaultOptions`, even though the provided function actually takes a different type. But as you say, :shrug:.